File: 2. Recursion/Medium and Hard/Palindrome-Partitioning.py
# Palindrome Partitioning (Leetcode)
# Leetcode Link : https://leetcode.com/problems/palindrome-partitioning/description/

# Time Complexity : O(2^n * n/2 * k)
# O(2^n) to generate every substring and O(n/2)  to check if the substring generated is a palindrome. O(k) is for inserting the palindromes in another data structure, where k  is the average length of the palindrome list.

# Space Complexity : O(k*x)
# k is the average length of the list having palindrome strings and x is the no. of lists having palindrome strings. The depth of the recursion tree is n, so the auxiliary space required is equal to the O(n).

class Solution:
    def isPalindrome(self, s):
        l = 0
        r = len(s) - 1

        while(l <= r):
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        
        return True

    def solve(self, s, ind, op, ans):
        n = len(s)
        if ind >= n:
            ans.append(op[:])
            return
        
        for i in range(ind, n):
            if self.isPalindrome(s[ind : i+1]):
                op.append(s[ind : i+1])
                self.solve(s, i + 1, op, ans)
                op.pop()

    def partition(self, s: str) -> List[List[str]]:
        ans = []
        self.solve(s, 0, [], ans)

        return ans

File: 2. Recursion/Backtracking/Palindrome-Partitioning.cpp
// Palindrome Partitioning (Leetcode)
// Leetcode Link : https://leetcode.com/problems/palindrome-partitioning/description/

// Time Complexity : O(n * 2^n)
// Space Complexity : O(n * 2^n)

class Solution {
public:
    // Function to check if a given string is a palindrome
    bool isPalindrome(string str){
        int i = 0;
        int j = str.size() - 1;

        while(i < j){
            if(str[i] != str[j]) return false;
            i++;
            j--;
        }

        return true;
    }

    void solve(string s, int start, vector<string> &palindromes, vector<vector<string>> &ans){
        // Base case: if start index reaches the end of the string
        if(start == s.size()){
            ans.push_back(palindromes); // Add the current partition to the answer
            return;
        }

        string substr; // Temporary string to hold substrings
        for(int i = start; i < s.size(); i++){
            substr += s[i]; // Form substring by adding one character at a time

            // Check if the substring is a palindrome
            if(isPalindrome(substr)){
                palindromes.push_back(substr); // Add palindrome to the list
                solve(s, i+1, palindromes, ans); // Recur for the rest of the string
                palindromes.pop_back(); // Backtrack to explore other partitions
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> palindromes;
        solve(s, 0, palindromes, ans);
        return ans;
    }
};